multiple target files loaded with file_sd

paulfantom opened the pull request:
Resolves #58

SuperQ commented on defaults/main.yml:
```diff
@@ -51,7 +52,7 @@ prometheus_scrape_configs:
   file_sd_configs:
   - files:
-    - "{{ prometheus_config_dir }}/file_sd/targets.yml"
+    - "{{ prometheus_config_dir }}/file_sd/*.yml"
```
This should not be a glob.

paulfantom replied:
Ok, but as I said in https://github.com/cloudalchemy/ansible-prometheus/issues/58#issuecomment-357447135, I don't see any other option. Users who want to tune prometheus to their custom setup will overwrite `prometheus_scrape_configs` and this glob won't be used. However for others this will be quite useful.

SuperQ replied:
It may be "useful", but it's considered an anti-pattern, and will encourage the wrong usage patterns.

paulfantom replied:
Ok, so it seems I need to write some crazy preflight check to notify (fail playbook) user when he provides wrong config.

SuperQ commented on tasks/configure.yml:
```diff
@@ -28,11 +28,12 @@
     group: prometheus
     mode: 0644
+  with_dict: "{{ prometheus_targets }}"
   when: prometheus_targets != []
```
This needs to be updated to `{}`

SuperQ commented on tasks/configure.yml:
```diff
@@ -28,11 +28,12 @@
 - name: configure prometheus static targets
   template:
```
Instead of a template, why not `copy` with `content: "item.value | to_nice_yaml(indent=2)"`

paulfantom replied:
This won't allow setting `{{ ansible_managed }}` comment as the first line.

SuperQ replied:
Would this work?

```yaml
content: |
  # {{ ansible_managed }}
  {{ item.value | to_nice_yaml(indent=2) }}
```

paulfantom replied:
No you can't. `ansible_managed` is available only in `template` module.
https://github.com/ansible/ansible/issues/34693
http://docs.ansible.com/ansible/latest/template_module.html#synopsis

SuperQ replied:
Bleh, ok, maybe something to fix in Ansible upstream, there are many cases where it would be nice if Ansible would simply write out a nice yaml file.

paulfantom replied:
Yeah, it could help in some situations. Go to https://github.com/ansible/ansible/issues/34693 and maybe we could change it into feature request :smile:

paulfantom commented:
Ok, now it should work. I've removed `*` and added one crazy preflight check based on:
- http://docs.ansible.com/ansible/latest/playbooks_filters.html#json-query-filter
- http://jmespath.org/specification.html#flatten-operator

This iterates over `prometheus_targets` dict keys and searches for target files in all `file_sd` sections of `promethus_scrape_configs`. It will fail on wrong configuration.

I think we should also add some info about it in README.md.

@SuperQ I couldn't find any info in official prometheus docs that using `*` in `file_sd_config` is an anti-pattern. On contrary I think current revision is even encouraging it's usage. Source: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#file_sd_config

SuperQ commented:
The anti-pattern is overloading `job` in the targets file itself.  The documentation is not encouraging the use, simply stating that it's possible.

The main reason I want to discourage the specific globbing, is that it encourages the job overloading, and makes it seem like you're supposed to do it that way.

In reality, the reason that you want multiple target files is so that you can update them asynchronously for each job.  When a target file updates, Prometheus triggers a reload of this file, no `HUP` is necessary.  But this reload causes all in-flight scrapes to be canceled.  This is OK for targets within a job, since that job is being updated.  But having all job targets in one file would cancel scrapes globally.

It also makes the configuration more confusing for beginners.

SuperQ commented on tasks/configure.yml:
```diff
@@ -28,11 +28,12 @@
-  when: prometheus_targets != []
+  with_dict: "{{ prometheus_targets }}"
+  when: prometheus_targets != {}
   notify:
```
This `notify` can be removed, target files are watched automatically.

paulfantom commented:
Thanks for explaining it :+1:

SuperQ commented on defaults/main.yml:
```diff
@@ -34,13 +34,14 @@ prometheus_global:
+#    - localhost:9100
+#    labels:
+#      env: test
+#      job: node
```
This `job` label example can be removed.

SuperQ commented:
One question about the defaults, why not un-comment and make the `node` job enabled by default?

paulfantom commented:
There is no good reason to have default values for `prometheus_targets` commented, so I un-commented them. This also reduced number of variables overwritten in test run.

SuperQ commented:
Awesome, :shipit:
